content/blog/2020-05-22-rt-template/index.md

```diff
@@ -0,0 +1,365 @@
+    Instruction *retInst = ReturnInst::Create(Ctx, pb);
+```
+This looks easy, but there's a caveat. Unlike the instructions, which are very general,
+now you have a specific function to replace. This turned out to be quite a challenge to
+extend to all the benchmarks in machsuite, even with a trivial function. (A function that
+reads precalculated values and fills in the arrays.) Not only the function signature has to be
+identical both in the llvm pass and the runtime library, the size of the arguments had to be the
+same size in the original program and the runtime function. 
+
+Therefore, functions offer a performant tradeoff for an accelerator. But in terms of productive 
+implementation, it is very program specific. The runtime library and the high-level 
+programmer needs to follow ad-hoc set of rules to use this template, even though it is
+possible to add some checks in the llvm pass with some engineering.
+
+## Replacing basic blocks
+Replacing a basic block can be a middle ground between these two granularities.
+A basic block can be considered as a collection of instructions, for the generation
+of interfaces. Since the interfaces for an instruction are clear, basic blocks will be an extension.
+It can be seen as a part of a function for the purposes of granularity. So a (loosely coupled) accelerator
+can work at a higher granularity than an instruction.
+
+```c
+    /* create the function call from runtime library */
+    LLVMContext& Ctx = F.getContext();
+    FunctionCallee rtFunc;
+    rtFunc = F.getParent()->getOrInsertFunction(
+      "bb_replace", 
+      Type::getInt32Ty(Ctx),
+      Type::getInt32Ty(Ctx)
+    );
+    // function call name and argument types have to be known
+
+    /* Add runtime basic block selection here */
+    
+    /* End of selection algorithm */
+    
+    /* Select a basic block */
+    BasicBlock* workList = NULL; // collects replaced basic block within iterator
+    /* mark basic blocks for deletion */
+    for (auto& B : F) {
+        for (auto& I : B) {
+            workList = &B;
+        }
+    }
+    
+    /* collect information on the basic block */
+    BasicBlock* predBB = workList->getSinglePredecessor();
+    BasicBlock* succBB = workList->getSingleSuccessor();
+    
+    /* delete basic block being replaced */
+    if (workList != NULL) {
+        workList->eraseFromParent();
+    }
+    
+    /* create new basic block */ 
+    BasicBlock *pb = BasicBlock::Create(Ctx, "simple", &F);
+    
+    /* create the runtime function call */
+    IRBuilder<> builder(pb);
+    Constant *i32_configs = ConstantInt::get(Type::getInt32Ty(Ctx), /*configuration*/, true);
+    Value* args[] = {i32_configs};
+    Value* ret =  builder.CreateCall(rtFunc, args);
+    
+    /* create a jump to next block */
+    Instruction *brInst = BranchInst::Create(succBB, pb);
+    
+    /* update jump from previous block */
+    Instruction *brPred = predBB->getTerminator();
+    brPred->setSuccessor(0, pb);
+```
```

Comment:
Again, we need an explanation for what the transformation actually does. What does it mean to replace a basic block with a function call? How is the terminator dealt with? How about the instruction operands and return values? What are the function arguments? And of course the big bugaboo: what about loads and stores—do those matter?

There are no wrong answers here—you just need to explain what you chose.